File: battle_real_time/lib/battle_real_time/contracts/consumers/player_events/pokemon_templates_list_contract.ex
defmodule BattleRealTime.Contracts.Consumers.PlayerEvents.PokemonTemplatesListContract do
  @moduledoc """
  Validates the payload for the `pokemon_templates_list` event on the `player_events` queue.
  """
  use BattleRealTime.Contracts.Contract

  @primary_key false
  embedded_schema do
    field(:player_id, :integer)
    field(:timestamp, :string)

    embeds_many :pokemon_templates, PokemonTemplateItem, primary_key: false do
      field(:id, :integer)
      field(:name, :string)
      field(:types, {:array, :string})
      field(:stats, :map)
      field(:sprite, :string)
    end
  end

  def changeset(struct, params) do
    struct
    |> cast(params, [:player_id, :timestamp])
    |> validate_required([:player_id])
    |> cast_embed(:pokemon_templates, with: &pokemon_template_changeset/2)
  end

  defp pokemon_template_changeset(struct, params) do
    struct
    |> cast(params, [:id, :name, :types, :stats, :sprite])
    |> validate_required([:id, :name, :types, :stats])
  end
end

File: battle_engine/app/contracts/publishers/pokemon_templates_list_contract.rb
# frozen_string_literal: true

module Contracts
  module Publishers
    class PokemonTemplatesListContract < Dry::Validation::Contract
      params do
        required(:player_id).filled(:integer)
        required(:pokemon_templates).array(:hash) do
          required(:id).filled(:integer)
          required(:name).filled(:string)
          required(:types).array(:string)
          required(:stats).filled(:hash)
          optional(:sprite).maybe(:string)
        end
        optional(:timestamp).maybe(:string)
      end
    end
  end
end
